File: app/src/main/java/com/absinthe/kage/device/cmd/HeartbeatCommand.java
package com.absinthe.kage.device.cmd;

import com.absinthe.kage.connect.protocol.IpMessageConst;
import com.absinthe.kage.connect.protocol.IpMessageProtocol;
import com.absinthe.kage.device.Command;
import com.absinthe.kage.device.CommandBuilder;
import com.absinthe.kage.device.client.Client;

import java.io.IOException;

public class HeartbeatCommand extends Command {

    public static final String HEARTBEAT_MESSAGE = "HEARTBEAT";
    public static final int LENGTH = 2;

    public HeartbeatCommand() {
        cmd = IpMessageConst.IS_ONLINE;
    }

    @Override
    public String pack() {
        return new CommandBuilder()
                .with(this)
                .append(HEARTBEAT_MESSAGE)
                .build();
    }

    @Override
    public void doWork(Client client, String received) {
        try {
            client.writeToStream(pack());
        } catch (IOException e) {
            e.printStackTrace();
            client.offline();
        }
    }

    @Override
    public boolean parseReceived(String received) {
        String[] splits = received.split(IpMessageProtocol.DELIMITER);
        if (splits.length == LENGTH) {
            try {
                return Integer.parseInt(splits[0]) == cmd;
            } catch (NumberFormatException e) {
                e.printStackTrace();
                return false;
            }
        } else {
            return false;
        }
    }
}

File: app/src/main/java/com/absinthe/kage/device/cmd/HeartbeatCommand.kt
package com.absinthe.kage.device.cmd

import com.absinthe.kage.connect.protocol.IpMessageConst
import com.absinthe.kage.connect.protocol.IpMessageProtocol
import com.absinthe.kage.device.Command
import com.absinthe.kage.device.CommandBuilder
import com.absinthe.kage.device.client.Client
import java.io.IOException

class HeartbeatCommand : Command() {

    override fun pack(): String {
        return CommandBuilder()
                .with(this)
                .append(HEARTBEAT_MESSAGE)
                .build()
    }

    override fun doWork(client: Client, received: String) {
        try {
            client.writeToStream(pack())
        } catch (e: IOException) {
            e.printStackTrace()
            client.offline()
        }
    }

    override fun parseReceived(received: String): Boolean {
        val splits = received.split(IpMessageProtocol.DELIMITER).toTypedArray()

        return if (splits.size == LENGTH) {
            try {
                splits[0].toInt() == cmd
            } catch (e: NumberFormatException) {
                e.printStackTrace()
                false
            }
        } else {
            false
        }
    }

    companion object {
        const val HEARTBEAT_MESSAGE = "HEARTBEAT"
        const val LENGTH = 2
    }

    init {
        cmd = IpMessageConst.IS_ONLINE
    }
}